### ISSUE TITLE
Typescript: missing type definitions.

### ISSUE THREAD

**Issue by ejez** [quasarframework/quasar#8642]
I am opening this issue to report missing types when I notice that.

Here is the first:
Cookie options type does not have a name, currently I'm doing:
```ts
Parameters<typeof Cookies.set>[2]
```
Which is a little ugly :disappointed: 

(Note: `sameSite` property can be improved: `'Lax' | 'Strict' | 'None'` instead of `string`)

Join me in making @IlCallo more busy :smile: 
Thanks @IlCallo for your hard work :1st_place_medal:

**Comment by ejez** [comment]
Few more ts issues:

1. When trying to access `Quasar.lang.props`:
`Property 'props' does not exist on type 'GlobalQuasarLanguage'.`

`props` defined here:
 https://github.com/quasarframework/quasar/blob/ed49443ff31647b1a379e87c9761384cb063ad4b/ui/src/lang.js#L78

2. Function `Quasar.lang.set` should allow a second argument (ssrContext)
https://next.quasar.dev/options/quasar-language-packs#dynamical-ssr-

Getting: `Expected 1 arguments, but got 2.` when trying to call it with ssrContext.

**Comment by IlCallo** [comment]
> When trying to access Quasar.lang.props:

Seems an internal API, it's not mentioned into docs AFAIK.
What is its use case?

> Function Quasar.lang.set should allow a second argument (ssrContext)

It should let you, if you have SSR feature flag file into your project (`src-ssr/ssr-feature-flag.d.ts`)
Otherwise it only accepts a parameter, since SSR isn't detected in your project

**Comment by ejez** [comment]
Thanks @IlCallo 
> What is its use case?

When the user changes the app language, I want to cache the original lang pack in case the user switches back to the original language and load the lang pack from a cached object instead of loading it again from the server.

>It should let you, if you have SSR feature flag file into your project 

I have the SSR feature flag enabled, I can find it in the boot file params for example.

**Comment by rstoenescu** [comment]
@ejez once you import a language pack, it's already there in a sort of "cache". Subsequent imports are not requiring network connection.

**Comment by ejez** [comment]
Thanks Razvan,
Thats the point, I am retrieving it from Quasar.lang.props, and reusing it later when needed, the issue is 'props' property is not present in 'Quasar' type. I am extending the type myself.

**Comment by ejez** [comment]
I am not talking about lang packs imported after the app has initialized, but the original pack the app was started with, the one defined in quasar.conf.js (or en-US if not set)

**Comment by ejez** [comment]
I am building an app ext for I18n, you can see where I am using 'props' here
https://github.com/ejez/i18n/blob/6fb9b4433af499f969054efa731ec8daca24fe9e/src/quasar-lang/index.ts#L11

(Extension in early development )

**Comment by ejez** [comment]
These are quick reproduction steps to clarify the issues.

```sh
quasar create lang-test --branch next

? Project name (internal usage for dev) lang-test
? Project product name (must start with letter if building mobile apps) Quasar App
? Project description A Quasar Framework app
? Pick your CSS preprocessor: SCSS
? Check the features needed for your project: ESLint (recommended), TypeScript
? Pick a component style: Composition
? Pick an ESLint preset: Prettier
? Continue to install project dependencies after the project has been created? (recommended) yarn

cd lang-test

quasar new boot -f ts q-lang
# Enable the boot file.
sed -i "/boot: \[/a \      'q-lang'" quasar.conf.js
# Enable Cookies plugin.
sed -i "s/plugins: \[/plugins: \[\n\        'Cookies'\n\      /" quasar.conf.js
# Split lang pack files in separate chunks.
sed -i "/return/a \    vendor: { remove: \['quasar/lang'\] }," quasar.conf.js
# Allow using ssr.
sed -i '/TODO vue3/,+9 s/^/\/\//' node_modules/@quasar/app/bin/quasar-dev node_modules/@quasar/app/bin/quasar-build
# The following will add the ssr feature flag.
quasar dev -m ssr
```

`src/boot/q-lang.ts`

```ts
import type { QuasarLanguage } from "quasar";
import type { QSsrContext } from "@quasar/app";
import { boot } from "quasar/wrappers";
import { Cookies, Quasar } from "quasar";

export default boot(async ({ ssrContext }) => {
  const cookies = process.env.SERVER
    ? Cookies.parseSSR(ssrContext as QSsrContext)
    : Cookies;

  // Check which lang is set.
  console.log(Quasar.lang.isoName); // => 'en-US'

  // Simulating existence of a previously set cookie.
  cookies.set("iso_name", "ro", { sameSite: "Lax" });

  // Load the lang pack corresponding to the cookie.
  let qLang = await importQLang(cookies.get("iso_name"));
  // The above line results in a network request.

  // Set Quasar lang.
  if (process.env.SERVER) {
    // @ts-expect-error function signature overload missing.
    Quasar.lang.set(qLang, ssrContext);
  } else {
    Quasar.lang.set(qLang);
  }

  // Check which lang is set.
  console.log(Quasar.lang.isoName); // => ro

  // After a while the user switches back to 'en-US'.

  // I was thinking that the following line will result in an un-necessary
  // network request, but you are right, the 'en-US' pack is already included in
  // the big app chunk, and won't be requested from the server.
  qLang = await importQLang("en-US");

  // Set Quasar lang.
  if (process.env.SERVER) {
    // @ts-expect-error function signature overload missing.
    Quasar.lang.set(qLang, ssrContext);
  } else {
    Quasar.lang.set(qLang);
  }

  // Check which lang is set.
  console.log(Quasar.lang.isoName); // => 'en-US'
});

async function importQLang(isoName: string) {
  return (<{ default: QuasarLanguage }>await import(
    /* webpackInclude: /(ro|en-US)\.js$/ */
    `quasar/lang/${isoName}`
  )).default;
}
```

So as mentioned in the above code snippet, I don't need `props` for my use case. Thanks @rstoenescu for hinting about it.

For the second issue, as you can see I needed to use `@ts-expect-error` to call `Quasar.lang.set` with a second argument, if not we get: `Expected 1 arguments, but got 2.`

Notice also the inconvenience of the need to cast `ssrContext` as `QSsrContext` before handing it over to `Cookies.parseSSR`, if not we get: `Argument of type 'QSsrContext | null | undefined' is not assignable to parameter of type 'LooseDictionary'.`

**Comment by ejez** [comment]
`CookiesGetMethodType` shoud include `null` in its return type as per the API (https://next.quasar.dev/quasar-plugins/cookies#api--cookies)

`ui/types/api/cookies.d.ts`

```diff
-export type CookiesGetMethodType = <T = string>(name: string) => T;
+export type CookiesGetMethodType = <T = string>(name: string) => T | null;
```

If not, someone could do:
```ts
const myString = Cookies.get('non_existing_cookie')
console.log(myString.toLowerCase())
```
and not get alerted (but the app will break at runtime).

(note however, the corresponding json `ui/src/plugins/Cookies.json` includes `null` in the return type.)

I wanted to submit a PR, but I was not sure if the types are autogenerated from json files, can you explain briefly @IlCallo how is it done.

**Comment by ejez** [comment]
> @ejez once you import a language pack, it's already there in a sort of "cache". Subsequent imports are not requiring network connection.

@rstoenescu I noticed that the 'en-US' pack gets bundled always un-necessarily, even if the app is not using it at all (`conf.framework.lang` set to another lang).

For now, the lang packs are relatively small, but if the lang pack model gets expanded later with additional strings, this won't be optimal.

I think `ui/src/lang.js` should be generated from a template (as it is done with files built from `app/templates`) with the correct language if set in `conf.framework.lang`.

**Comment by IlCallo** [comment]
@ejez can you contact me privately on Discord, if you still have some time to fix some of these TS problems?

**Comment by rstoenescu** [comment]
Fix will be available in Quasar 2.3.3
